gcp_variant_transforms/libs/vcf_file_composer.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2018 Google Inc.  All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Composes multiple files in GCS to one VCF file."""
+
+import multiprocessing
+from typing import Iterable, List  # pylint: disable=unused-import
+
+from apache_beam.io import filesystems
+from apache_beam.io.gcp import gcsio
+
+from google.cloud import storage
+
+_MAX_NUM_OF_BLOBS_PER_COMPOSE = 32
+
+
+def compose_vcf_data_files(project,  # type: str
+                           vcf_data_files_folder,  # type: str
+                           output_file,  # type: str
+                           delete=True  # type: bool
+                          ):
+  # type: (...) -> None
+  """Composes multiple VCF data files to one VCF data file.
+
+  It composes VCF data shards to one VCF data file and deletes the original VCF
+  data files if `delete` is True.
+
+  Args:
+    project: The project name.
+    vcf_data_files_folder: The folder that contains all VCF data files.
+    output_file: The final VCF file path.
+    delete: If true, delete the original VCF data files.
+  """
+  bucket_name, blob_prefix = gcsio.parse_gcs_path(vcf_data_files_folder)
+  client = storage.Client(project)
+  bucket = client.get_bucket(bucket_name)
+  multi_process_composer = MultiProcessComposer(project, bucket_name,
+                                                blob_prefix)
+  composed_vcf_data_file = multi_process_composer.get_composed_blob()
+  output_file_blob = _create_blob(client, output_file)
+  output_file_blob.rewrite(composed_vcf_data_file)
+  if delete:
+    bucket.delete_blobs(bucket.list_blobs(prefix=blob_prefix))
+
+
+def _compose_files(project, bucket_name, blob_names, composite_name):
+  # type: (str, str, List[str], str) -> None
+  """Composes multiple files (up to 32 objects) in GCS to one.
+
+  Args:
+    project: The project name.
+    bucket_name: The name of the bucket where the `components` and the new
+      composite are saved.
+    blob_names: A list of blob object names.
+    composite_name: Name of the new composite.
+  """
+  bucket = storage.Client(project).get_bucket(bucket_name)
+  output_file_blob = bucket.blob(composite_name)
+  output_file_blob.content_type = 'text/plain'
+  blobs = [bucket.get_blob(blob_name) for blob_name in blob_names]
+  output_file_blob.compose(blobs)
```

Comment:
@allieychen This is nice, but where are you doing any integrity checks?

Reply:
No problem Allie, that's fine - it was only to ensure the tests have might not be sprinkled with false positives.  One other thing to keep in mind is some sort of retry policy - funny things with high-throughput data can sneak up on you unexpectedly even when chunking at the max of 32 :)